File: src/main/java/ac/dnd/server/admission/domain/model/EventsData.java
package ac.dnd.server.admission.domain.model;

import java.time.LocalDateTime;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;

import ac.dnd.server.admission.exception.EventNotFoundException;

public record EventsData(
	List<EventData> values
) {

	public EventData findCurrentActiveEvent(
		final LocalDateTime now
	) {
		return findCompletedEvent()
			.orElseGet(() -> findNearestPendingEvent(now)
				.orElseThrow(EventNotFoundException::new));
	}

	private Optional<EventData> findCompletedEvent() {
		return this.values.stream()
			.filter(EventData::isCompleted)
			.findFirst();
	}

	private Optional<EventData> findNearestPendingEvent(final LocalDateTime now) {
		return this.values.stream()
			.filter(event -> event.isAnnounceable(now))
			.min(Comparator.comparing(EventData::resultAnnouncementDateTime));
	}
}

File: src/main/kotlin/ac/dnd/server/admission/domain/model/EventsData.kt
package ac.dnd.server.admission.domain.model

import ac.dnd.server.admission.exception.EventNotFoundException
import java.time.LocalDateTime

data class EventsData(
    val values: List<EventData>
) {
    fun findCurrentActiveEvent(now: LocalDateTime): EventData {
        return findCompletedEvent()
            ?: findNearestPendingEvent(now)
            ?: throw EventNotFoundException()
    }

    private fun findCompletedEvent(): EventData? {
        return values.firstOrNull { it.isCompleted() }
    }

    private fun findNearestPendingEvent(now: LocalDateTime): EventData? {
        return values
            .filter { it.isAnnounceable(now) }
            .minByOrNull { it.resultAnnouncementDateTime }
    }
}